### ISSUE TITLE
[BUG] Backing image manager fails when SELinux is enabled

### ISSUE THREAD

**Issue by ejweber** [longhorn/longhorn#6108]
## Describe the bug (🐛 if you encounter this issue)

When SELinux is enabled, backing-image-manager pods end up in a crash loop.

## To Reproduce

1. Deploy an RKE2 cluster on Rocky 9.2. Do not disable SELinux (`getenforce` returns enforcing).
2. Run the Longhorn integration tests. (There is no doubt a much simpler recreate, but I haven't investigated this much yet.)
3. Observe that instead of being marked as FAILED, tests are marked as ERROR. This indicates something broken in the cluster itself (not just a test case failure).
4. Observe the below symptoms:

With `kubectl` access:

```
[rocky@ip-10-0-1-71 ~]$ kubectl get pod -n longhorn-system | grep backing
backing-image-manager-1050-1130                     0/1     Error               0          3s
backing-image-manager-1050-237b                     0/1     Error               0          3s
backing-image-manager-1050-b9da                     0/1     ContainerCreating   0          1s

[rocky@ip-10-0-1-71 ~]$ kubectl logs -n longhorn-system backing-image-manager-1050-1130
time="2023-06-12T21:55:45Z" level=fatal msg="Error running start command" error="cannot find disk config file /data/longhorn-disk.cfg: open /data/longhorn-disk.cfg: permission denied"
```

On a worker node:

```
[rocky@ip-10-0-2-113 ~]$ sudo ausearch -m AVC -ts recent
----
time->Mon Jun 12 21:56:24 2023
type=PROCTITLE msg=audit(1686606984.732:9363): proctitle=6261636B696E672D696D6167652D6D616E61676572002D2D6465627567006461656D6F6E002D2D6C697374656E00302E302E302E303A38303030002D2D73796E632D6C697374656E00302E302E302E303A38303031
type=SYSCALL msg=audit(1686606984.732:9363): arch=c000003e syscall=257 success=no exit=-13 a0=ffffffffffffff9c a1=c0001a07b0 a2=80000 a3=0 items=0 ppid=192769 pid=192885 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="backing-image-m" exe="/usr/local/bin/backing-image-manager" subj=system_u:system_r:container_t:s0:c576,c979 key=(null)
type=AVC msg=audit(1686606984.732:9363): avc:  denied  { read } for  pid=192885 comm="backing-image-m" name="longhorn-disk.cfg" dev="xvda5" ino=134481951 scontext=system_u:system_r:container_t:s0:c576,c979 tcontext=system_u:object_r:container_var_lib_t:s0 tclass=file permissive=0
```

## Expected behavior

Backing image manager works fine.

## Log or Support bundle

https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4086

## Environment

 - Longhorn version: v1.5.0-rc1
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: v1.27.2+rke2
   - Number of management node in the cluster: 1
   - Number of worker node in the cluster: 3
 - Node config
   - OS type and version: Rocky v9.2
   - CPU per node:
   - Memory per node:
   - Disk type(e.g. SSD/NVMe):
   - Network bandwidth between the nodes:
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): AWS
 - Number of Longhorn volumes in the cluster:

## Additional context

I THINK this is how https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4086/console is failing / failed. I will try to confirm when it is complete and I have a support bundle.

**Comment by innobead** [comment]
This should be day 1 issue. cc @longhorn/qa

**Comment by innobead** [comment]
cc @ChanYiLin

**Comment by ejweber** [comment]
> I THINK this is how https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4086/console is failing / failed. I will try to confirm when it is complete and I have a support bundle.

There were two different issues causing this behavior. I am fixing one in https://github.com/longhorn/longhorn-tests/pull/1424, but this one causes similar behavior.

**Comment by ejweber** [comment]
For some reason, this issue is NOT present when using k3s. It is only reproduceable when using RKE2.

**Comment by innobead** [comment]
Is it probably related to the policies installed differently by both distros?

**Comment by ejweber** [comment]
I now understand what is causing this and am considering how to fix it.

## Why did we only see it in RKE2 and not k3s?

This was a false negative. Even though the k3s install script includes `container-selinux`, SELinux must be explicitly enabled by some k3s configuration option. Without doing so, all k3s pods run with the `spc_t` (super privileged container) context, giving them extra, unexpected permissions. When SELinux is explicitly enabled, k3s fails in the exact what RKE2 does.

## What is the root cause?

Here is the more complete audit log output:

```
*****  Plugin catchall (100. confidence) suggests   **************************

If you believe that backing-image-manager should be allowed read access on the longhorn-disk.cfg file by default.
Then you should report this as a bug.
You can generate a local policy module to allow this access.
Do
allow this access for now by executing:
# ausearch -c 'backing-image-m' --raw | audit2allow -M my-backingimagem
# semodule -X 300 -i my-backingimagem.pp


Additional Information:
Source Context                system_u:system_r:container_t:s0:c174,c229
Target Context                system_u:object_r:container_var_lib_t:s0
Target Objects                longhorn-disk.cfg [ file ]
Source                        backing-image-m
Source Path                   /usr/local/bin/backing-image-manager
Port                          <Unknown>
Host                          ip-10-0-2-129.us-east-2.compute.internal
Source RPM Packages
Target RPM Packages
SELinux Policy RPM            selinux-policy-targeted-38.1.11-2.el9_2.2.noarch
Local Policy RPM              selinux-policy-targeted-38.1.11-2.el9_2.2.noarch
Selinux Enabled               True
Policy Type                   targeted
Enforcing Mode                Enforcing
Host Name                     ip-10-0-2-129.us-east-2.compute.internal
Platform                      Linux ip-10-0-2-129.us-east-2.compute.internal
                              5.14.0-284.11.1.el9_2.x86_64 #1 SMP
                              PREEMPT_DYNAMIC Tue May 9 17:09:15 UTC 2023 x86_64
                              x86_64
Alert Count                   1
First Seen                    2023-06-15 14:40:44 UTC
Last Seen                     2023-06-15 14:40:44 UTC
Local ID                      381261be-0236-4a99-97fb-9b1360b22566

Raw Audit Messages
type=AVC msg=audit(1686840044.560:914114): avc:  denied  { read } for  pid=3401056 comm="backing-image-m" name="longhorn-disk.cfg" dev="xvda5" ino=76871079 scontext=system_u:system_r:container_t:s0:c174,c229 tcontext=system_u:object_r:container_var_lib_t:s0 tclass=file permissive=0


type=SYSCALL msg=audit(1686840044.560:914114): arch=x86_64 syscall=openat success=no exit=EACCES a0=ffffffffffffff9c a1=c0000407e0 a2=80000 a3=0 items=0 ppid=3401004 pid=3401056 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=backing-image-m exe=/usr/local/bin/backing-image-manager subj=system_u:system_r:container_t:s0:c174,c229 key=(null)

Hash: backing-image-m,container_t,container_var_lib_t,file,read
```

Because it is running from an unprivileged container, `backing-image-manager` has context `container_t`. Because the directory containing all Longhorn data was created by `containerd`, it and all of its children have context `container_var_lib_t`. The rule that ensures this looks like the following (though this part is a bit hazy).

```
[rocky@ip-10-0-2-129 lib]$ sesearch -T -s container_runtime_t -t var_lib_t
type_transition container_runtime_t var_lib_t:dir container_var_lib_t;
```

There is no rule to allow a process running as `container_t` to read `container_var_lib_t` files. On the other hand, there is such a rule for `spc_t`.

```
[rocky@ip-10-0-2-129 lib]$ sesearch -A -s container_t -t container_var_lib_t -c file -p read

[rocky@ip-10-0-2-129 lib]$ sesearch -A -s spc_t -t container_var_lib_t -c file -p read
allow files_unconfined_type file_type:file { append audit_access create execute execute_no_trans getattr ioctl link lock map mounton open quotaon read relabelfrom relabelto rename setattr swapon unlink watch watch_mount watch_reads watch_sb watch_with_perm write };
```

## What do we do?

There are a number of very specific contexts that `container_t` is allowed to read from, but none of them are particularly useful, as the container runtime will be creating the `/var/lib/longhorn/` directory for us. 

In the long term, we can consider writing our own SELinux policy for Longhorn (though it will be difficult to deliver, as usually this is done in an RPM via a package manager).

In the short term, if we want `backing-image-manager` to run successfully on Longhorn, we will have to run it privileged.

**Comment by innobead** [comment]
This is a good analysis.

The diskPath is created by longhorn-manager via ns with contained, so this is a permission dependency. Since the longhorn manager is privileged, so does the backing image manager. This is the right fix for me. 

[/controller/backing_image_manager_controller.go#L862-L871](https://github.com/longhorn/longhorn-manager/blob/ed4e17678762f2fb6cd62e96eedd4be4cf3cb7e1/controller/backing_image_manager_controller.go#L862-L871)
```go
			Volumes: []v1.Volume{
				{
					Name: "disk-path",
					VolumeSource: v1.VolumeSource{
						HostPath: &v1.HostPathVolumeSource{
							Path: bim.Spec.DiskPath,
						},
					},
				},
			},
``` 

For longhorn-specific selinux policies, we could think about that if really needed in the future. However, the below ns selection would be a concern to impact what policies we should. 

[/vendor/github.com/longhorn/go-iscsi-helper/util/process.go#L55-L65](https://github.com/longhorn/longhorn-manager/blob/ed4e17678762f2fb6cd62e96eedd4be4cf3cb7e1/vendor/github.com/longhorn/go-iscsi-helper/util/process.go#L55-L65)
```go
func GetHostNamespacePath(hostProcPath string) string {
	pf := NewProcessFinder(hostProcPath)
	containerNames := []string{DockerdProcess, ContainerdProcess, ContainerdProcessShim}
	for _, name := range containerNames {
		proc, err := pf.FindAncestorByName(name)
		if err == nil {
			return fmt.Sprintf("%s/%d/ns/", hostProcPath, proc.Pid)
		}
	}
	return fmt.Sprintf("%s/%d/ns/", hostProcPath, 1)
}

```

**Comment by innobead** [comment]
This also means the backing image function is not working from day 1 on SELinux enforced environment. @khushboo-rancher we need to review SELinux test case to cover this. 

cc @longhorn/qa

**Comment by innobead** [comment]
@ejweber please remember to update the issue status accordingly when working on the issue. 

Need to backport to 1.4 and 1.3.

**Comment by innobead** [comment]
@yangchiu Please help verify this to see if the issue has been fixed in the master and 1.5.

**Comment by chriscchien** [comment]
Verified **pass** on 
- longhorn master (longhorn-manager `a3b16b`)
- longhorn v1.5.x (longhorn-manager `d85f96`)

In a Rocky Linux 9.2. rke2 cluster, perform` test_backing_image.py`, all test case passed on [master](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4254/) and [v1.5.x](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4251/) pipeline
In a Oracle Linux 8.6 k3s cluster, perform `test_backing_image.py`, all test case passed on master and v1.5.x as below

master
![image](https://github.com/longhorn/longhorn/assets/73337386/5d6599a2-db4c-4cb4-bae7-5f92d4d7089a)

v.1.5.x
![image](https://github.com/longhorn/longhorn/assets/73337386/f3234129-f834-4b57-ae01-b797d2b499f3)
